### PR TITLE
Fix from_pretrained with corrupted state_dict

### PR DESCRIPTION
**Pull request by sgugger**
# What does this PR do?

As we discovered in #12843, when a state dictionary contains keys for the body of the model that are not prefixed *and* keys for the head, the body is loaded but the head is ignored with no warning.

This PR fixes that by keeping track of the expected key that do not contain the prefix and erroring out if we load only the body of the model and there are some keys to load in that list of expected keys that do not contain the prefix. I chose the error since those kinds of state dictionaries should not exist, since `from_pretrained` or `torch.save(model.state_dict())` do not generate those.

### CODE REVIEW

**Comment by sgugger** [comment]
The test caught something weird with `sshleifer/tiny-distilbert-base-uncased-finetuned-sst-2-english`.

This model is used in the benchmark tests and in the zero shot pipeline but that model is beyond salvation: its weights have the name of BERT when it's a DistilBERT architecture, the labels of the config don't match the weights. Loading it just results in a random model.

File: src/transformers/modeling_utils.py [hunk 1]
```diff
@@ -1490,6 +1491,11 @@ def load(module: nn.Module, prefix=""):
             start_prefix = cls.base_model_prefix + "."
         if hasattr(model, cls.base_model_prefix) and not has_prefix_module:
             model_to_load = getattr(model, cls.base_model_prefix)
+            if any(key in expected_keys_not_prefixed for key in loaded_keys) > 0:
```
Comment (patrickvonplaten) [inline src/transformers/modeling_utils.py:1494]: nit - why not just `if any(key in expected_keys_not_prefixed for key in loaded_keys):` ?

File: src/transformers/modeling_utils.py [hunk 2]
```diff
@@ -1490,6 +1491,11 @@ def load(module: nn.Module, prefix=""):
             start_prefix = cls.base_model_prefix + "."
         if hasattr(model, cls.base_model_prefix) and not has_prefix_module:
```
Comment (patrickvonplaten) [inline src/transformers/modeling_utils.py:1492]: Not really part of this PR, but I think in general we could/should do some refactoring here:

1)
`hasattr(model, cls.base_model_prefix) and not has_prefix_module`  is actually the same as `remove_prefix` no? 

`model` is the randomly loaded model and only has a `cls.base_model_prefix` if `expects_prefix_module` is True I think

2) We define `prefix` further above, but then keep using `cls.base_model_prefix`. 

3) The `model_to_load` variable is confusing, we should just replace the line: 
`model_to_load = getattr(model, cls.base_model_prefix)` below with `model = getattr(model, prefix)` IMO and remove `model_to_load` completely

Comment (patrickvonplaten) [reply src/transformers/modeling_utils.py:1492]: Regarding 1) I think it was actually me that defined the `remove_prefix` variable. Thinking about it again the name is not very good and doesn't help the reader to understand what's going on. Think a better name would be something like `load_base_into_head_model`

**Review by patrickvonplaten — APPROVED** [review]
Totally agree with the PR! 

Not really related to the PR, but it took me a lot of time to read through the code to understand what's going on => think we should/could do some refactoring here. The naming of the variables is not great and in general we define to many new variables here IMO

File: src/transformers/modeling_utils.py (hunk 1 above)
Comment (sgugger) [reply src/transformers/modeling_utils.py:1494]: Woopsy, had a `len` before that `any` and forgot to remove the comparison.

File: tests/test_benchmark.py [hunk 3]
```diff
@@ -49,7 +49,7 @@ def test_inference_no_configs(self):
         self.check_results_dict_not_empty(results.memory_inference_result)
 
     def test_inference_no_configs_only_pretrain(self):
-        MODEL_ID = "sshleifer/tiny-distilbert-base-uncased-finetuned-sst-2-english"
+        MODEL_ID = "sgugger/tiny-distilbert-classification"
```
Comment (LysandreJik) [inline tests/test_benchmark.py:52]: Would be nice to eventually move that to a model on `hf-internal-testing`; I'll upload the tokenizers shortly so that we may have tiny model from that org exclusively.

**Review by LysandreJik — APPROVED** [review]
Thank you for the change @sgugger, LGTM!

File: tests/test_benchmark.py (hunk 3 above)
Comment (sgugger) [reply tests/test_benchmark.py:52]: Yes, that was quick fix to be able to merge this PR quickly.
